### PR TITLE
Air Drop Mechanism Control and Maestro Class

### PR DESCRIPTION
**Pull request by Ooglely**
Creates a Maestro class in order to be able to control the servos that will be on the boards, and uses it to run the functions to use the air drop mechanism.

### CODE REVIEW

**Comment by Ooglely** [comment]
Waiting on a working power cable so that I can test the class functions.

File: flight/maestro/air_drop.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+"""
+File containing the functions to control dropping the bottles from the drone
+"""
+
+import maestro
```
Comment (fallscameron01) [inline flight/maestro/air_drop.py]: Imports should be relative to root

File: flight/maestro/air_drop.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+BOARD: maestro.Maestro = maestro.Maestro()
```
Comment (fallscameron01) [inline flight/maestro/air_drop.py]: Don't think this should be a constant. Alternative would be to create a class that encompasses this and the other functions in this file.

File: flight/maestro/maestro.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    Attributes
+    ----------
+    serial: serial.Serial
+        Serial object to be able to read and send messages to and from the board.
+
+    command_prefix: str
+        The command protocol prefix that is sent before every command to the board.
+
+    targets: list(int)
+        A list of the current targets for each channel.
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Shouldn't have extra newlines between these.
```suggestion
    serial: serial.Serial
        Serial object to be able to read and send messages to and from the board.
    command_prefix: str
        The command protocol prefix that is sent before every command to the board.
    targets: list(int)
        A list of the current targets for each channel.
```

File: flight/maestro/maestro.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+    Methods
+    -------
```
Comment (fallscameron01) [inline flight/maestro/maestro.py:23]: Same thing for your methods

File: flight/maestro/maestro.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        self,
+        connection_address: str = "/dev/ttyACM0",
+        device_number: int = 12,
+        baud_rate: int = 9600,
+    ):
+        """
+        Initializes a serial object to connect to the Maestro board
+
+        Parameters
+        ----------
```
Comment (fallscameron01) [inline flight/maestro/maestro.py:58]: And in all the other docstrings

File: flight/maestro/maestro.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        connection_address: str
+            The command serial port of the Maestro board (usually '/dev/ttyACM0')
+
+        device_number: int
+            The device number that is used to address the board in Pololu Protocol commands
+
+        baud_rate: int
+            The rate of communication configured for the board (default 9600)
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: State the defaults for these in the docstring

File: flight/maestro/maestro.py [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        """
+        Read a 16 bit response from the Maestro board and convert it to an integer
+
+        Returns
+        -------
+        result: int
+            The response from the board
+        """
+        lsb = ord(self.serial.read())
+        msb = ord(self.serial.read())
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hints for these 2

File: flight/maestro/maestro.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        target: int
+            The target position in units of quarter-microseconds
+
+        Notes
+        -----
+        To set the servos on the drone to full left, set the target to 4000.
+        To set the servos to full right, set the target to 8000.
+        The range of the servo target is 4000 - 8000.
+        """
+        low, high = self._get_bits(target)
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hints for theses 2 variables.

File: flight/maestro/maestro.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        command = chr(0x04) + chr(channel) + chr(low) + chr(high)
+
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: And this one needs a type hint as well

File: flight/maestro/maestro.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        Parameters
+        ----------
+        channel: int
+            The channel of the desired servo
+
+        speed: int
+            The speed limit to set the channel to in units of quarter-microseconds/10 ms
+        """
+        low, high = self._get_bits(speed)
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hints

File: flight/maestro/maestro.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+        command = chr(0x07) + chr(channel) + chr(low) + chr(high)
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hint

File: flight/maestro/maestro.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            The target acceleration in units of quarter-microseconds/10 ms/80 ms
+
+        Notes
+        -----
+        The range for acceleration is from 0 to 255.
+        0 indicates no acceleration limit.
+        """
+        low, high = self._get_bits(accel)
+
+        command = chr(0x09) + chr(channel) + chr(low) + chr(high)
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: And here

File: flight/maestro/maestro.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+            The channel number of the desired servo
+
+        Returns
+        -------
+        position: int
+            The current position of the servo in quarter-microseconds
+        """
+        command = chr(0x10) + chr(channel)
+        self._send_command(command)
+        result = self._read_bits()
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hints for `command` and `result`

File: flight/maestro/maestro.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+        print(result)
+        return result
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Shouldn't be printing

File: flight/maestro/maestro.py [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+        -------
+        response: bool
+            True if servos are still moving, false if not
+
+        Notes
+        -----
+        Only works with Maestro 12, 18, 24
+        """
+        self._send_command(chr(0x13))
+        response = self.serial.read()
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hint for `response`

File: flight/maestro/maestro.py [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        return result
+
+    def get_moving_state(self) -> bool:
+        """
+        States if any servos are currently moving
+
+        Returns
+        -------
+        response: bool
+            True if servos are still moving, false if not
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]:
```suggestion
        response: bool
            True if servos are still moving, False if not
```

File: flight/maestro/maestro.py [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+        """
+        Returns the error code if there are any
+
+        Returns
+        -------
+        error: int
+            The error code
+        """
+        self._send_command(chr(0x21))
+        result = self._read_bits()
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Type hint for `result`

File: flight/maestro/maestro.py [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+        print(result)
+        return result
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: remove print statement

**Review by fallscameron01 — CHANGES_REQUESTED** [review]
Few changes to make, mostly around documentation and type hints.

File: flight/maestro/air_drop.py (hunk 2 above)
Comment (cjhr95) [reply flight/maestro/air_drop.py]: Another alternative would be to pass a Maestro class into each function, similar to the drone object we pass for flight stuff

File: flight/maestro/maestro.py [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+
+    Attributes
+    ----------
+    serial: serial.Serial
+        Serial object to be able to read and send messages to and from the board.
+
+    command_prefix: str
+        The command protocol prefix that is sent before every command to the board.
+
+    targets: list(int)
```
Comment (cjhr95) [inline flight/maestro/maestro.py]: type hint for the list should be `targets: list[int]`

File: flight/maestro/maestro.py [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+        ----------
+        command: str
+            The command (in bits) to be sent to the board
+        """
+        self.serial.write(bytes(self.command_prefix + command, "latin-1"))
+
+    def _get_bits(self, integer: int) -> tuple[int, int]:
+        """
+        Converts an integer value to the 7 high bit, 7 low bit format that the Pololu commands use
+        See: https://www.pololu.com/docs/0J40/5.e
```
Comment (cjhr95) [inline flight/maestro/maestro.py]: Throw this link into a separate "Notes" section in your docstring

**Review by cjhr95 — CHANGES_REQUESTED** [review]
I don't believe the functions should be async, it might mess up the servo functionality, just when they're called in the state we can run it asynchronously.

**Review by cjhr95 — DISMISSED** [review]
Seems solid to me

File: flight/maestro/maestro.py [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+
+        Returns
+        -------
+        position: int
+            The current position of the servo in quarter-microseconds
+        """
+        command: str = chr(0x10) + chr(channel)
+        self._send_command(command)
+        result: int = self._read_bits()
+        return result
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: The return variable is `result`

File: flight/maestro/maestro.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+        Returns the error code if there are any
+
+        Returns
+        -------
+        error: int
+            The error code
+        """
+        self._send_command(chr(0x21))
+        result: int = self._read_bits()
+        return result
```
Comment (fallscameron01) [inline flight/maestro/maestro.py]: Same here

**Review by fallscameron01 — CHANGES_REQUESTED** [review]
Couple of docstring issues

**Review by fallscameron01 — APPROVED** [review]
Looks good

**Review by cjhr95 — APPROVED** [review]
Kinda rude to dismiss my review but whatever

**Pull request merged**